pkg/cloud/services/ec2/launchtemplate_test.go

```diff
@@ -309,3 +436,850 @@ func TestService_LaunchTemplateNeedsUpdate(t *testing.T) {
+
+			mpScope, err := setupMachinePoolScope(client, mockScope)
+			if err != nil {
+				t.Fatalf("failed while creating machinepool scope with err: %+v", err)
+			}
+
+			s := &Service{
+				scope:     mockScope,
+				EC2Client: mockEC2Client,
+				SSMClient: mockSSMClient,
+			}
+
+			tc.expect(mockScope.EXPECT(), mockEC2Client.EXPECT())
+			tc.check(s.CreateLaunchTemplateVersion(mpScope, aws.String("imageID"), userData))
+		})
+	}
+}
+
+func TestBuildLaunchTemplateTagSpecificationRequest(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name   string
+		expect func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder)
+		check  func(m []*ec2.LaunchTemplateTagSpecificationRequest)
+	}{
+		{
+			name: "failed during tag specification request for instance",
+			expect: func(scope *mocks.MockEC2ScopeMockRecorder, m *mock_ec2iface.MockEC2APIMockRecorder) {
+				scope.AdditionalTags().Return(infrav1.Tags{})
+				scope.Name().Return("cluster-name").AnyTimes()
+			},
+			check: func(res []*ec2.LaunchTemplateTagSpecificationRequest) {
+				expected := []*ec2.LaunchTemplateTagSpecificationRequest{
+					{
+						ResourceType: aws.String(ec2.ResourceTypeInstance),
+						Tags:         defaultEC2Tags("aws-mp-name", "cluster-name"),
+					},
+					{
+						ResourceType: aws.String(ec2.ResourceTypeVolume),
+						Tags:         defaultEC2Tags("aws-mp-name", "cluster-name"),
+					},
+				}
+				// sorting tags for comparing each request tags during reflect.DeepEqual()
+				for _, each := range res {
+					sortTags(each.Tags)
+				}
+				if !reflect.DeepEqual(res, expected) {
+					t.Fatalf("Expected len of both should be same")
+				}
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			scheme, err := setupScheme()
+			if err != nil {
+				t.Fatalf("failed during scheme registration")
+			}
+
+			client := fake.NewClientBuilder().WithScheme(scheme).Build()
+
+			mockScope := mocks.NewMockEC2Scope(mockCtrl)
+			mockEC2Client := mock_ec2iface.NewMockEC2API(mockCtrl)
+			mockSSMClient := mock_ssmiface.NewMockSSMAPI(mockCtrl)
+
+			mpScope, err := setupMachinePoolScope(client, mockScope)
+			if err != nil {
+				t.Fatalf("failed while creating machinepool scope with err: %+v", err)
+			}
+
+			s := &Service{
+				scope:     mockScope,
+				EC2Client: mockEC2Client,
+				SSMClient: mockSSMClient,
+			}
+			tc.expect(mockScope.EXPECT(), mockEC2Client.EXPECT())
+			tc.check(s.buildLaunchTemplateTagSpecificationRequest(mpScope))
+		})
+	}
+}
+
+func TestDiscoverLaunchTemplateAMI(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	testCases := []struct {
+		name              string
+		awsLaunchTemplate expinfrav1.AWSLaunchTemplate
+		machineTemplate   clusterv1.MachineTemplateSpec
+		expect            func(m *mock_ec2iface.MockEC2APIMockRecorder)
+		check             func(*string, error)
+	}{
+		{
```

Comment:
Is it possible to dd following scenarios here:
- Should fetch imageLookupFormat if not present in launchtemplate
- Should fetch ImageLookupOrg if not present in launchtemplate
- Should fetch ImageLookupBaseOS if not present in launchtemplate
- Should do eks EMI lookup in case of EKSManaged cluster(you can think of covering eksAMILookup() func here with different combinations to cover all edge cases)

Reply:
done